File: Leetcode/Sliding Window/Subarray_with_given_sum.py
# Question asked in Amazon Facebook Google Visa
# Find Subarray in the given sum (GeeksForGeeks POTD - )

#User function Template for python3

#Function to find a continuous sub-array which adds up to a given number.
class Solution:
    def subArraySum(self, arr, n, s):
       #Write your code here
        i = 0
        j = 0
        add = arr[0]
        if s == 0:
            return [-1]
        if s == arr[0]:
           return [1, 1]
        else:
            while (i < n and j < n):
                if add == s:
                    return [i+1, j+1]
                elif add < s:
                    if j+1 < n:
                        j += 1
                        add += arr[j]
                    else:
                        return [-1]

                else:
                    add -= arr[i]
                    i += 1
        return [-1]

File: Leetcode/Sliding Window/Subarray_with_given_sum.cpp
// Question asked in Amazon Facebook Google Visa
// Find Subarray in the given sum (GeeksForGeeks POTD - dated 19 Aug, 2023)

// Driver Code Starts

#include <bits/stdc++.h>
using namespace std;

// Driver Code Ends

class Solution
{
    public:
    //Function to find a continuous sub-array which adds up to a given number.
    vector<int> subarraySum(vector<int>arr, int n, long long s)
    {
        // Your code here
        int left = 0;
        int right = 0;
        vector<int> ans;
        
        if (s == 0) {
            ans.push_back(-1);
            return ans;
        }
        
        bool isFound = false;
        int sum = arr[0];
        
        while (right < n) {
            if (sum == s) {
                isFound = true;
                break;
            }
            else if (sum < s) {
                right++;
                if (right < n) sum += arr[right];
            } else {
                sum -= arr[left];
                left++; 
            }
        }
        if (isFound) {
            ans.push_back(left + 1);
            ans.push_back(right + 1);
            return ans;
        }
        ans.push_back(-1);
        return ans;
    }
};